File: t/crypto_basis.py
import hashlib
import hmac
import secrets
import base64
from typing import Callable, List, Dict, Any, Tuple
from dataclasses import dataclass


@dataclass
class CryptoTest:
    name: str
    func: Callable
    inputs: List[Any]
    expected: Any


class CryptoBasis:
	KNOWN_PATTERNS = {
		"identity": b"\x00" * 32,
		"all_ones": b"\xff" * 32,
		"ascii_a": b"a" * 32,
		"ascii_0": b"0" * 32,
		"null_byte": b"\x00",
	}
	
	TestResult = Tuple[str, bool, str, Any, Any]
	
	@staticmethod
	def sha256(data: bytes) -> bytes:
		return hashlib.sha256(data).digest()
	
	@staticmethod
	def sha512(data: bytes) -> bytes:
		return hashlib.sha512(data).digest()
	
	@staticmethod
	def hmac_sha256(key: bytes, data: bytes) -> bytes:
		return hmac.new(key, data, hashlib.sha256).digest()
	
	@staticmethod
	def hmac_sha512(key: bytes, data: bytes) -> bytes:
		return hmac.new(key, data, hashlib.sha512).digest()
	
	@staticmethod
	def random_bytes(n: int) -> bytes:
		return secrets.token_bytes(n)
	
	@staticmethod
	def random_hex(n: int) -> str:
		return secrets.token_hex(n)
	
	@staticmethod
	def pbkdf2(password: bytes, salt: bytes, iterations: int, keylen: int) -> bytes:
		return hashlib.pbkdf2_hmac("sha256", password, salt, iterations, dklen=keylen)
	
	@staticmethod
	def scrypt(password: bytes, salt: bytes, n: int, r: int, p: int, keylen: int) -> bytes:
		return hashlib.scrypt(password, salt=salt, n=n, r=r, p=p, dklen=keylen)
	
	@staticmethod
	def md5(data: bytes) -> bytes:
		return hashlib.md5(data).digest()
	
	@staticmethod
	def sha1(data: bytes) -> bytes:
		return hashlib.sha1(data).digest()
	
	@staticmethod
	def sha384(data: bytes) -> bytes:
		return hashlib.sha384(data).digest()
	
	@staticmethod
	def sha3_256(data: bytes) -> bytes:
		return hashlib.sha3_256(data).digest()
	
	@staticmethod
	def sha3_512(data: bytes) -> bytes:
		return hashlib.sha3_512(data).digest()
	
	@staticmethod
	def blake2b(data: bytes, digest_size: int = 64) -> bytes:
		return hashlib.blake2b(data, digest_size=digest_size).digest()
	
	@staticmethod
	def blake2s(data: bytes, digest_size: int = 32) -> bytes:
		return hashlib.blake2s(data, digest_size=digest_size).digest()


class CryptoTestHarness:
	BASIS_CASES = [
		"empty_input",
		"single_byte",
		"short_string",
		"long_input",
		"binary_null",
		"repetition",
		"boundary_values",
		"random_input",
	]
	
	def __init__(self):
		self.results: List[CryptoBasis.TestResult] = []
		self.test_vectors: Dict[str, bytes] = {}
		self._build_test_vectors()
	
	def _build_test_vectors(self):
		self.test_vectors = {
			"empty_input": b"",
			"single_byte": b"x",
			"short_string": b"hello world",
			"long_input": b"a" * 10000,
			"binary_null": b"\x00\x01\x02\x00",
			"repetition": b"\x55" * 256,
			"boundary_values": bytes(range(256)),
			"random_input": CryptoBasis.random_bytes(1024),
		}
	
	def _register_result(self, test_name: str, passed: bool, msg: str, expected: Any, actual: Any):
		self.results.append((test_name, passed, msg, expected, actual))
	
	def test_hash_function(self, name: str, func: Callable[[bytes], bytes]) -> bool:
		all_passed = True
		for case_name, input_data in self.test_vectors.items():
			try:
				result = func(input_data)
				if result is None or len(result) == 0:
					self._register_result(f"{name}_{case_name}", False, "empty result", "non-empty", result)
					all_passed = False
				else:
					self._register_result(f"{name}_{case_name}", True, "ok", len(result), len(result))
			except Exception as e:
				self._register_result(f"{name}_{case_name}", False, str(e), "no error", str(e))
				all_passed = False
		return all_passed
	
	def test_hmac_function(self, name: str, func: Callable[[bytes, bytes], bytes]) -> bool:
		all_passed = True
		key = CryptoBasis.random_bytes(32)
		for case_name, input_data in self.test_vectors.items():
			try:
				result = func(key, input_data)
				if result is None or len(result) == 0:
					self._register_result(f"{name}_{case_name}", False, "empty result", "non-empty", result)
					all_passed = False
				else:
					self._register_result(f"{name}_{case_name}", True, "ok", len(result), len(result))
			except Exception as e:
				self._register_result(f"{name}_{case_name}", False, str(e), "no error", str(e))
				all_passed = False
		return all_passed
	
	def test_deterministic(self, func: Callable[[bytes], bytes], input_name: str = "test") -> bool:
		input_data = self.test_vectors[input_name]
		r1 = func(input_data)
		r2 = func(input_data)
		is_equal = r1 == r2
		self._register_result(f"deterministic_{input_name}", is_equal, "consistent output", r1, r2)
		return is_equal
	
	def test_collision_resistance(self, func: Callable[[bytes], bytes], sample_count: int = 100) -> bool:
		hashes = set()
		all_unique = True
		for i in range(sample_count):
			data = f"test_{i}".encode()
			h = func(data)
			if h in hashes:
				all_unique = False
				break
			hashes.add(h)
		self._register_result("collision_resistance", all_unique, f"{sample_count} unique", "unique", all_unique)
		return all_unique
	
	def test_output_lengths(self, func: Callable[[bytes], bytes], expected_lengths: List[int]) -> bool:
		all_passed = True
		input_data = b"test"
		result = func(input_data)
		if len(result) not in expected_lengths:
			self._register_result("output_length", False, f"got {len(result)}", expected_lengths, len(result))
			return False
		self._register_result("output_length", True, "ok", expected_lengths, len(result))
		return True
	
	def test_known_answer(self, func: Callable, input_data: bytes, expected: bytes, test_name: str) -> bool:
		result = func(input_data)
		is_equal = result == expected
		self._register_result(test_name, is_equal, "known answer", expected.hex(), result.hex())
		return is_equal
	
	def run_all_tests(self) -> Dict[str, Any]:
		passed = 0
		failed = 0
		
		print("=" * 60)
		print("PYTHON CRYPTOGRAPHY TEST HARNESS")
		print("=" * 60)
		
		self.test_hash_function("sha256", CryptoBasis.sha256)
		self.test_hash_function("sha512", CryptoBasis.sha512)
		self.test_hash_function("md5", CryptoBasis.md5)
		self.test_hash_function("sha1", CryptoBasis.sha1)
		self.test_hash_function("sha384", CryptoBasis.sha384)
		self.test_hash_function("sha3_256", CryptoBasis.sha3_256)
		self.test_hash_function("sha3_512", CryptoBasis.sha3_512)
		self.test_hash_function("blake2b", CryptoBasis.blake2b)
		self.test_hash_function("blake2s", CryptoBasis.blake2s)
		
		self.test_hmac_function("hmac_sha256", CryptoBasis.hmac_sha256)
		self.test_hmac_function("hmac_sha512", CryptoBasis.hmac_sha512)
		
		self.test_deterministic(CryptoBasis.sha256, "empty_input")
		self.test_deterministic(CryptoBasis.sha256, "single_byte")
		self.test_deterministic(CryptoBasis.sha256, "short_string")
		
		self.test_collision_resistance(CryptoBasis.sha256)
		
		self.test_output_lengths(CryptoBasis.sha256, [32])
		self.test_output_lengths(CryptoBasis.sha512, [64])
		
		known_answers = [
			("sha256", CryptoBasis.sha256, b"", "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"),
			("sha256", CryptoBasis.sha256, b"abc", "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad".encode()),
			("sha256", CryptoBasis.sha256, b"hello world", "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9".encode()),
		]
		for name, func, inp, exp in known_answers:
			self.test_known_answer(func, inp, exp, f"kat_{name}")
		
		for name, passed_val, msg, exp, act in self.results:
			status = "PASS" if passed_val else "FAIL"
			print(f"[{status}] {name}: {msg}")
			if passed_val:
				passed += 1
			else:
				failed += 1
		
		print("=" * 60)
		print(f"RESULTS: {passed} passed, {failed} failed")
		print("=" * 60)
		
		return {"passed": passed, "failed": failed, "total": passed + failed}


def main():
	harness = CryptoTestHarness()
	results = harness.run_all_tests()
	
	if results["failed"] > 0:
		print(f"\nCRITICAL: {results['failed']} tests failed!")
		return 1
	print("\nAll crypto basis tests passed!")
	return 0


if __name__ == "__main__":
	exit(main())

File: t/crypto_basis.go
package main

import (
	"crypto"
	"crypto/md5"
	"crypto/rand"
	"crypto/sha1"
	"crypto/sha256"
	"crypto/sha512"
	"crypto/blake2b"
	"crypto/blake2s"
	"crypto/hmac"
	"crypto/subtle"
	"encoding/hex"
	"fmt"
	"hash"
	"math/big"
	"time"
)

type TestResult struct {
	name    string
	passed bool
	msg    string
	expect interface{}
	actual interface{}
}

type CryptoBasis struct{}

type CryptoTestHarness struct {
	results     []TestResult
	testVectors map[string][]byte
}

func NewCryptoTestHarness() *CryptoTestHarness {
	return &CryptoTestHarness{
		results: make([]TestResult, 0),
		testVectors: map[string][]byte{
			"empty_input":    []byte{},
			"single_byte":   []byte("x"),
			"short_string":  []byte("hello world"),
			"long_input":   make([]byte, 10000),
			"binary_null":  []byte{0x00, 0x01, 0x02, 0x00},
			"repetition":   make([]byte, 256),
			"boundary_values": make([]byte, 256),
			"random_input": make([]byte, 1024),
		},
	}
}

func init() {
	for i := range make([]int, 10000) {
		_NewCryptoTestHarness().testVectors["long_input"][i] = 'a'
	}
	for i := 0; i < 256; i++ {
		_NewCryptoTestHarness().testVectors["repetition'][i] = 0x55
	}
	for i := 0; i < 256; i++ {
		_NewCryptoTestHarness().testVectors["boundary_values'][i] = byte(i)
	}
	rand.Read(_NewCryptoTestHarness().testVectors["random_input"])
}

func (c *CryptoBasis) SHA256(data []byte) []byte {
	h := sha256.Sum256(data)
	return h[:]
}

func (c *CryptoBasis) SHA512(data []byte) []byte {
	h := sha512.Sum512(data)
	return h[:]
}

func (c *CryptoBasis) HMAC_SHA256(key, data []byte) []byte {
	m := hmac.New(sha256.New, key)
	m.Write(data)
	return m.Sum(nil)
}

func (c *CryptoBasis) HMAC_SHA512(key, data []byte) []byte {
	m := hmac.New(sha512.New, key)
	m.Write(data)
	return m.Sum(nil)
}

func (c *CryptoBasis) MD5(data []byte) []byte {
	h := md5.Sum(data)
	return h[:]
}

func (c *CryptoBasis) SHA1(data []byte) []byte {
	h := sha1.Sum(data)
	return h[:]
}

func (c *CryptoBasis) SHA384(data []byte) []byte {
	h := sha512.Sum384(data)
	return h[:]
}

func (c *CryptoBasis) SHA3_256(data []byte) []byte {
	return nil
}

func (c *CryptoBasis) SHA3_512(data []byte) []byte {
	return nil
}

func (c *CryptoBasis) BLAKE2b(data []byte, digestSize int) []byte {
	if digestSize == 0 {
		digestSize = 64
	}
	h, _ := blake2b.New(digestSize, nil)
	h.Write(data)
	return h.Sum(nil)
}

func (c *CryptoBasis) BLAKE2s(data []byte, digestSize int) []byte {
	if digestSize == 0 {
		digestSize = 32
	}
	h, _ := blake2s.New(digestSize, nil)
	h.Write(data)
	return h.Sum(nil)
}

func (c *CryptoBasis) RandomBytes(n int) []byte {
	b := make([]byte, n)
	rand.Read(b)
	return b
}

func (h *CryptoTestHarness) registerResult(name string, passed bool, msg string, expect, actual interface{}) {
	h.results = append(h.results, TestResult{name, passed, msg, expect, actual})
}

func (h *CryptoTestHarness) testHashFunction(name string, fn func([]byte) []byte) bool {
	allPassed := true
	for caseName, inputData := range h.testVectors {
		result := fn(inputData)
		if result == nil || len(result) == 0 {
			h.registerResult(name+"_"+caseName, false, "empty result", "non-empty", result)
			allPassed = false
		} else {
			h.registerResult(name+"_"+caseName, true, "ok", len(result), len(result))
		}
	}
	return allPassed
}

func (h *CryptoTestHarness) testHmacFunction(name string, fn func(key, data []byte) []byte) bool {
	allPassed := true
	key := CryptoBasis{}.RandomBytes(32)
	for caseName, inputData := range h.testVectors {
		result := fn(key, inputData)
		if result == nil || len(result) == 0 {
			h.registerResult(name+"_"+caseName, false, "empty result", "non-empty", result)
			allPassed = false
		} else {
			h.registerResult(name+"_"+caseName, true, "ok", len(result), len(result))
		}
	}
	return allPassed
}

func (h *CryptoTestHarness) testDeterministic(fn func([]byte) []byte, inputName string) bool {
	inputData := h.testVectors[inputName]
	r1 := fn(inputData)
	r2 := fn(inputData)
	isEqual := subtle.ConstantTimeCompare(r1, r2) == 1
	h.registerResult("deterministic_"+inputName, isEqual, "consistent output", hex.EncodeToString(r1), hex.EncodeToString(r2))
	return isEqual
}

func (h *CryptoTestHarness) testCollisionResistance(fn func([]byte) []byte, sampleCount int) bool {
	hashes := make(map[string]bool)
	allUnique := true
	for i := 0; i < sampleCount; i++ {
		data := []byte(fmt.Sprintf("test_%d", i))
		h := hex.EncodeToString(fn(data))
		if hashes[h] {
			allUnique = false
			break
		}
		hashes[h] = true
	}
	h.registerResult("collision_resistance", allUnique, fmt.Sprintf("%d unique", sampleCount), "unique", allUnique)
	return allUnique
}

func (h *CryptoTestHarness) testOutputLengths(fn func([]byte) []byte, expectedLengths []int) bool {
	inputData := []byte("test")
	result := fn(inputData)
	found := false
	for _, l := range expectedLengths {
		if len(result) == l {
			found = true
			break
		}
	}
	if !found {
		h.registerResult("output_length", false, fmt.Sprintf("got %d", len(result)), expectedLengths, len(result))
		return false
	}
	h.registerResult("output_length", true, "ok", expectedLengths, len(result))
	return true
}

func (h *CryptoTestHarness) testKnownAnswer(fn func([]byte) []byte, inputData, expected []byte, testName string) bool {
	result := fn(inputData)
	isEqual := subtle.ConstantTimeCompare(result, expected) == 1
	h.registerResult(testName, isEqual, "known answer", hex.EncodeToString(expected), hex.EncodeToString(result))
	return isEqual
}

func (h *CryptoTestHarness) runAllTests() (int, int, int) {
	passed := 0
	failed := 0

	fmt.Println("============================================================")
	fmt.Println("GO CRYPTOGRAPHY TEST HARNESS")
	fmt.Println("============================================================")

	c := CryptoBasis{}

	h.testHashFunction("sha256", c.SHA256)
	h.testHashFunction("sha512", c.SHA512)
	h.testHashFunction("md5", c.MD5)
	h.testHashFunction("sha1", c.SHA1)
	h.testHashFunction("sha384", c.SHA384)
	h.testHashFunction("blake2b", func(d []byte) []byte { return c.BLAKE2b(d, 64) })
	h.testHashFunction("blake2s", func(d []byte) []byte { return c.BLAKE2s(d, 32) })

	h.testHmacFunction("hmac_sha256", c.HMAC_SHA256)
	h.testHmacFunction("hmac_sha512", c.HMAC_SHA512)

	h.testDeterministic(c.SHA256, "empty_input")
	h.testDeterministic(c.SHA256, "single_byte")
	h.testDeterministic(c.SHA256, "short_string")

	h.testCollisionResistance(c.SHA256, 100)

	h.testOutputLengths(c.SHA256, []int{32})
	h.testOutputLengths(c.SHA512, []int{64})

	knownAnswers := []struct {
		name   string
		fn    func([]byte) []byte
		input []byte
		expect []byte
	}{
		{"sha256", c.SHA256, []byte{}, decodeHex("e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855")},
		{"sha256", c.SHA256, []byte("abc"), decodeHex("ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad")},
		{"sha256", c.SHA256, []byte("hello world"), decodeHex("b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9")},
	}

	for _, kat := range knownAnswers {
		h.testKnownAnswer(kat.fn, kat.input, kat.expect, "kat_"+kat.name)
	}

	for _, r := range h.results {
		status := "PASS"
		if !r.passed {
			status = "FAIL"
			failed++
		} else {
			passed++
		}
		fmt.Printf("[%s] %s: %s\n", status, r.name, r.msg)
	}

	fmt.Println("============================================================")
	fmt.Printf("RESULTS: %d passed, %d failed\n", passed, failed)
	fmt.Println("============================================================")

	return passed, failed, passed + failed
}

func decodeHex(s string) []byte {
	b, _ := hex.DecodeString(s)
	return b
}

func main() {
	harness := NewCryptoTestHarness()
	passed, failed, _ := harness.runAllTests()

	if failed > 0 {
		fmt.Printf("\nCRITICAL: %d tests failed!\n", failed)
	} else {
		fmt.Println("\nAll crypto basis tests passed!")
	}

	if failed > 0 {
		time.Sleep(time.Second)
	}
}